network/p2p/network_test.go

```diff
@@ -5,309 +5,245 @@ package p2p
-	nodeID := ids.GenerateTestNodeID()
-	chainID := ids.GenerateTestID()
+	require := require.New(t)
+	ctx := context.Background()
 
-	ctxKey := new(string)
-	ctxVal := new(string)
-	*ctxKey = "foo"
-	*ctxVal = "bar"
+	sender := FakeSender{
+		SentAppRequest: make(chan []byte, 1),
+	}
+	network := NewNetwork(logging.NoLog{}, sender, prometheus.NewRegistry(), "")
 
-	tests := []struct {
-		name        string
-		requestFunc func(t *testing.T, network *Network, client *Client, sender *common.SenderTest, handler *mocks.MockHandler, wg *sync.WaitGroup)
-	}{
-		{
-			name: "app request",
-			requestFunc: func(t *testing.T, network *Network, client *Client, sender *common.SenderTest, handler *mocks.MockHandler, wg *sync.WaitGroup) {
-				sender.SendAppRequestF = func(ctx context.Context, nodeIDs set.Set[ids.NodeID], requestID uint32, request []byte) error {
-					for range nodeIDs {
-						go func() {
-							require.NoError(t, network.AppRequest(ctx, nodeID, requestID, time.Time{}, request))
-						}()
-					}
+	client, err := network.NewAppProtocol(handlerID, &NoOpHandler{})
+	require.NoError(err)
 
-					return nil
-				}
-				sender.SendAppResponseF = func(ctx context.Context, _ ids.NodeID, requestID uint32, response []byte) error {
-					go func() {
-						ctx = context.WithValue(ctx, ctxKey, ctxVal)
-						require.NoError(t, network.AppResponse(ctx, nodeID, requestID, response))
-					}()
+	wantResponse := []byte("response")
+	wantNodeID := ids.GenerateTestNodeID()
+	done := make(chan struct{})
 
-					return nil
-				}
-				handler.EXPECT().
-					AppRequest(context.Background(), nodeID, gomock.Any(), request).
-					DoAndReturn(func(context.Context, ids.NodeID, time.Time, []byte) ([]byte, error) {
-						return response, nil
-					})
-
-				callback := func(ctx context.Context, actualNodeID ids.NodeID, actualResponse []byte, err error) {
-					defer wg.Done()
-
-					require.NoError(t, err)
-					require.Equal(t, ctxVal, ctx.Value(ctxKey))
-					require.Equal(t, nodeID, actualNodeID)
-					require.Equal(t, response, actualResponse)
-				}
+	callback := func(_ context.Context, gotNodeID ids.NodeID, gotResponse []byte, err error) {
+		require.Equal(wantNodeID, gotNodeID)
+		require.NoError(err)
+		require.Equal(wantResponse, gotResponse)
 
-				require.NoError(t, client.AppRequestAny(context.Background(), request, callback))
-			},
-		},
-		{
-			name: "app request failed",
-			requestFunc: func(t *testing.T, network *Network, client *Client, sender *common.SenderTest, handler *mocks.MockHandler, wg *sync.WaitGroup) {
-				sender.SendAppRequestF = func(ctx context.Context, nodeIDs set.Set[ids.NodeID], requestID uint32, request []byte) error {
-					for range nodeIDs {
-						go func() {
-							require.NoError(t, network.AppRequestFailed(ctx, nodeID, requestID))
-						}()
-					}
+		close(done)
+	}
 
-					return nil
-				}
+	require.NoError(client.AppRequest(ctx, set.Of(wantNodeID), []byte("request"), callback))
+	<-sender.SentAppRequest
```

Comment:
Should we assert what we're sending here?

Reply:
Same for other reads on `sender` channels below

Reply:
my two cents: the current level of testing seems fine to me. Test seems to be focused around callbacks (as per comment on top); moreover bytes actually sent are prefixed and asserting their comment seems kind of testing an implementation detail to me.

Reply:
I added an assertion here sine the full request/response flow is checked

Reply:
> I added an assertion here

Which assertion do you mean?

Reply:
See `TestClientPrefixesMessages`, made a specific test for this. Forgot to update comment here.